src/mqtt_client/mod.rs

```diff
@@ -199,11 +230,69 @@ where
             mqtt,
             state: sm::StateMachine::new(sm::Context::new(clock)),
             settings,
-            settings_prefix,
             prefix: String::from(prefix),
+            listing_state: None,
+            properties_cache: None,
+            pending_response: None,
         })
     }
 
+    fn handle_listing(&mut self) {
+        let Some(ref mut iter) = &mut self.listing_state else {
+            return;
+        };
+
+        let Some(ref props) = self.properties_cache else {
+            return
+        };
+
+        let props = minimq::types::Properties::DataBlock(props);
+
+        let mut data = [0u8; MESSAGE_SIZE];
+
+        while self.mqtt.client().can_publish(QoS::AtLeastOnce) {
+            // Note(unwrap): Publishing should not fail because `can_publish()` was checked before
+            // attempting this publish.
+            let response: Response<MAX_TOPIC_LENGTH> = iter
+                .next()
+                .map(|path| Response::custom(ResponseCode::Continue, &path))
+                .unwrap_or_else(Response::ok);
+
+            let message = match serde_json_core::to_slice(&response, &mut data) {
+                Ok(len) => &data[..len],
+
+                _ => {
+                    let response: Response<32> = Response::error("Buffer too small");
+                    let Ok(len) = serde_json_core::to_slice(&response, &mut data) else {
+                        log::warn!("Message size is too small for responses");
```

Comment:
Honestly, I don't think we need the level of detail here. We'll get a line number and unique message for users to refer back to. They won't have any clue what the "buffer too small" response is and I would think it would just obfuscate what they need to do (i.e. increase the message size)